### PR TITLE
Task 3 to Master

### PR DESCRIPTION
**Pull request by Gaidukevich-Kirill**
task3 to master

### CODE REVIEW

File: Task3/Task3/Task3/Boxes/Box.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                if (item.GetType() == figure.GetType() && item.Side == figure.Side)
+                {
+                    alreadyExist = BoxOfFigures.Contains(item);
+                }
+            }
+
+            if (alreadyExist == false)
+            {
+                throw new AddException("This figure is already in the box");
+            }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Boxes/Box.cs]: I don't understand this logic.
Why do we set already exiast to true?

File: Task3/Task3/Task3/Boxes/Box.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            else
+            {
+                BoxOfFigures.Add(item);
+            }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Boxes/Box.cs]: Can we move the action out of the else block?

File: Task3/Task3/Task3/Boxes/Box.cs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            if (sameExist == false)
+            {
+                throw new CompareException("Figure isn't found.");
+            }
+            else
+            {
+                Console.WriteLine(
+                    $"The list has equivalent figure: {BoxOfFigures[position]}, " +
+                    $"with side: {BoxOfFigures[position].Side}");
+            }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Boxes/Box.cs]: Can we move the action out of the else block?

File: Task3/Task3/Task3/Figures/Figure.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                throw new PaintException("Cannot paint this figure.");
+            }
+
+            if (IsTypePaper.CheckType(this.GetType()))
+            {
+                if (this.ColorFigure != Color.WithoutColor)
+                {
+                    throw new PaintException("Cannot paint this figure.");
+                }
+            }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/Figure.cs]: Can we union this two if?

File: Task3/Task3/Task3/Figures/Figure.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+        public abstract double Perimeter
+        {
+            get;
+        }
+
+        public abstract double Area
+        {
+            get;
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/Figure.cs]: Use simple property construction without 'enter' => ` public abstract double Area { get; }`

File: Task3/Task3/Task3/Figures/FilmCircle.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public FilmCircle(double radius)
+        {
+            Side = radius;
+        }
+
+        public FilmCircle(ref FilmEquilateralTriangle triangle, double side)
+        {
+            if (side <= triangle.Side / (2 * Math.Sqrt(3)))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/FilmCircle.cs]: Can we move this logic to another private method?

File: Task3/Task3/Task3/Figures/FilmCircle.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            if (side <= square.Side / 2)
+            {
+                Side = side;
+                square = null;
+            }
+            else
+            {
+                throw new CarveException("It is impossible to carve this figure.");
+            }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/FilmCircle.cs]: Move counting logic to another method

File: Task3/Task3/Task3/Figures/FilmCircle.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public override double Area
+        {
+            get
+            {
+                return Math.PI * Side * Side;
+            }
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/FilmCircle.cs]: We can use next property: `public override double Area => Math.PI * Side * Side;`
The same for Perimeter

File: Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public FilmEquilateralTriangle(double side)
+        {
+            Side = side;
+        }
+
+        public FilmEquilateralTriangle(ref FilmCircle circle, double side)
+        {
+            if (side <= circle.Side * Math.Sqrt(3))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs]: Move logic to another method.

File: Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            else
+            {
+                throw new CarveException("It is impossible to carve this figure.");
+            }
+        }
+
+        public FilmEquilateralTriangle(ref FilmSquare square, double side)
+        {
+            if (side * Math.Sqrt(3) / 2 <= square.Side * Math.Sqrt(2))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs]: Move logic to another method.

File: Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        public override double Area
+        {
+            get
+            {
+                return Side * Side * Math.Sqrt(3) / 4;
+            }
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs]: See another comment about property

File: Task3/Task3/Task3/Figures/FilmSquare.cs [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public FilmSquare(double side)
+        {
+            Side = side;
+        }
+
+        public FilmSquare(ref FilmEquilateralTriangle triangle, double side)
+        {
+            if (side <= triangle.Side / (1 + 2 / Math.Sqrt(3)))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/FilmSquare.cs]: The same

File: Task3/Task3/Task3/Figures/FilmSquare.cs [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            else
+            {
+                throw new CarveException("It is impossible to carve this figure.");
+            }
+        }
+
+        public FilmSquare(ref FilmCircle circle, double side)
+        {
+            if (side * Math.Sqrt(2) <= circle.Side * 2)
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/FilmSquare.cs]: The same

File: Task3/Task3/Task3/Figures/FilmSquare.cs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        public override double Area
+        {
+            get
+            {
+                return Side * Side;
+            }
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/FilmSquare.cs]: The same

File: Task3/Task3/Task3/Figures/PaperCircle.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public PaperCircle(double radius)
+        {
+            Side = radius;
+        }
+
+        public PaperCircle(ref PaperEquilateralTriangle triangle, double side)
+        {
+            if (side <= triangle.Side / (2 * Math.Sqrt(3)))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PaperCircle.cs]: The same

File: Task3/Task3/Task3/Figures/PaperCircle.cs [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            else
+            {
+                throw new CarveException("It is impossible to carve this figure.");
+            }
+        }
+
+        public PaperCircle(ref PaperSquare square, double side)
+        {
+            if (side <= square.Side / 2)
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PaperCircle.cs]: The same

File: Task3/Task3/Task3/Figures/PaperCircle.cs [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        public override double Area
+        {
+            get
+            {
+                return Math.PI * Side * Side;
+            }
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PaperCircle.cs]: The same

File: Task3/Task3/Task3/Figures/PaperEquilateralTriangle.cs [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public PaperEquilateralTriangle(double side)
+        {
+            Side = side;
+        }
+
+        public PaperEquilateralTriangle(ref PaperCircle circle, double side)
+        {
+            if (side <= circle.Side * Math.Sqrt(3))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PaperEquilateralTriangle.cs]: The same

File: Task3/Task3/Task3/Figures/PaperEquilateralTriangle.cs [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            else
+            {
+                throw new CarveException("It is impossible to carve this figure.");
+            }
+        }
+
+        public PaperEquilateralTriangle(ref PaperSquare square, double side)
+        {
+            if (side * Math.Sqrt(3) / 2 <= square.Side * Math.Sqrt(2))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PaperEquilateralTriangle.cs]: The same

File: Task3/Task3/Task3/Figures/PaperEquilateralTriangle.cs [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        public override double Area
+        {
+            get
+            {
+                return Side * Side * Math.Sqrt(3) / 4;
+            }
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PaperEquilateralTriangle.cs]: The same

File: Task3/Task3/Task3/Figures/PaperSquare.cs [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public PaperSquare(double side)
+        {
+            Side = side;
+        }
+
+        public PaperSquare(ref PaperEquilateralTriangle triangle, double side)
+        {
+            if (side <= triangle.Side / (1 + 2 / Math.Sqrt(3)))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PaperSquare.cs]: The same

File: Task3/Task3/Task3/Figures/PaperSquare.cs [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            else
+            {
+                throw new CarveException("It is impossible to carve this figure.");
+            }
+        }
+
+        public PaperSquare(ref PaperCircle circle, double side)
+        {
+            if (side * Math.Sqrt(2) <= circle.Side * 2)
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PaperSquare.cs]: The same

File: Task3/Task3/Task3/Figures/PaperSquare.cs [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        public override double Area
+        {
+            get
+            {
+                return Side * Side;
+            }
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PaperSquare.cs]: The same

File: Task3/Task3/Task3/Figures/PlasticCircle.cs [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public PlasticCircle(double radius)
+        {
+            Side = radius;
+        }
+
+        public PlasticCircle(ref PlasticEquilateralTriangle triangle, double side)
+        {
+            if (side <= triangle.Side / (2 * Math.Sqrt(3)))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PlasticCircle.cs]: The same

File: Task3/Task3/Task3/Figures/PlasticCircle.cs [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            else
+            {
+                throw new CarveException("It is impossible to carve this figure.");
+            }
+        }
+
+        public PlasticCircle(ref PlasticSquare square, double side)
+        {
+            if (side <= square.Side / 2)
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PlasticCircle.cs]: The same

File: Task3/Task3/Task3/Figures/PlasticCircle.cs [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        public override double Area
+        {
+            get
+            {
+                return Math.PI * Side * Side;
+            }
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PlasticCircle.cs]: The same

File: Task3/Task3/Task3/Figures/PlasticEquilateralTriangle.cs [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public PlasticEquilateralTriangle(double side)
+        {
+            Side = side;
+        }
+
+        public PlasticEquilateralTriangle(ref PlasticCircle circle, double side)
+        {
+            if (side <= circle.Side * Math.Sqrt(3))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PlasticEquilateralTriangle.cs]: The same

File: Task3/Task3/Task3/Figures/PlasticEquilateralTriangle.cs [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            else
+            {
+                throw new CarveException("It is impossible to carve this figure.");
+            }
+        }
+
+        public PlasticEquilateralTriangle(ref PlasticSquare square, double side)
+        {
+            if (side * Math.Sqrt(3) / 2 <= square.Side * Math.Sqrt(2))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PlasticEquilateralTriangle.cs]: The same

File: Task3/Task3/Task3/Figures/PlasticEquilateralTriangle.cs [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public override double Area
+        {
+            get
+            {
+                return Side * Side * Math.Sqrt(3) / 4;
+            }
+        }
+    }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PlasticEquilateralTriangle.cs]: The same

File: Task3/Task3/Task3/Figures/PlasticSquare.cs [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public PlasticSquare(double side)
+        {
+            Side = side;
+        }
+
+        public PlasticSquare(ref PlasticEquilateralTriangle triangle, double side)
+        {
+            if (side <= triangle.Side / (1 + 2 / Math.Sqrt(3)))
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PlasticSquare.cs]: The same

File: Task3/Task3/Task3/Figures/PlasticSquare.cs [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            else
+            {
+                throw new CarveException("It is impossible to carve this figure.");
+            }
+        }
+
+        public PlasticSquare(ref PlasticCircle circle, double side)
+        {
+            if (side * Math.Sqrt(2) <= circle.Side * 2)
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PlasticSquare.cs]: The same

File: Task3/Task3/Task3/Figures/PlasticSquare.cs [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        public override double Area
+        {
+            get
+            {
+                return Side * Side;
+            }
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/Figures/PlasticSquare.cs]: The same

File: Task3/Task3/Task3/FileManagment/XmlBoxFileWorker.cs [hunk 33]
```diff
@@ -0,0 +1,184 @@
...
+                }
+            }
+
+            using (FileStream fs = new FileStream(path, FileMode.OpenOrCreate))
+            {
+                xmlSerializer.Serialize(fs, boxPlasticBox);
+
+                Console.WriteLine("Object has been serialized");
+            }
+        }
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/FileManagment/XmlBoxFileWorker.cs]: Can we create single methods for this operations?

File: Task3/Task3/Task3/FileManagment/XmlBoxFileWorker.cs [hunk 34]
```diff
@@ -0,0 +1,184 @@
...
+            {
+                if (IsTypePlastic.CheckType(figure.GetType()))
+                {
+                    boxPlasticFigures.AddFigure(figure);
+                }
+            }
+
+            StreamWriter myWriter = new StreamWriter(path);
+            mySerializer.Serialize(myWriter, boxPlasticFigures);
+            myWriter.Close();
```
Comment (Shumilov-Denis) [inline Task3/Task3/Task3/FileManagment/XmlBoxFileWorker.cs]: Can we create single methods for this operations?

**Review by Shumilov-Denis — CHANGES_REQUESTED** [review]

File: Task3/Task3/Task3/Boxes/Box.cs (hunk 1 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Boxes/Box.cs]: fixed

File: Task3/Task3/Task3/Boxes/Box.cs (hunk 2 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Boxes/Box.cs]: fixed

File: Task3/Task3/Task3/Boxes/Box.cs (hunk 3 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Boxes/Box.cs]: fixed

File: Task3/Task3/Task3/Figures/Figure.cs (hunk 4 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/Figure.cs]: fixed

File: Task3/Task3/Task3/Figures/Figure.cs (hunk 5 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/Figure.cs]: fixed

File: Task3/Task3/Task3/Figures/FilmCircle.cs (hunk 6 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/FilmCircle.cs]: fixed

File: Task3/Task3/Task3/Figures/FilmCircle.cs (hunk 7 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/FilmCircle.cs]: fixed

File: Task3/Task3/Task3/Figures/FilmCircle.cs (hunk 8 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/FilmCircle.cs]: fixed

File: Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs (hunk 9 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs]: fixed

File: Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs (hunk 10 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs]: fixed

File: Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs (hunk 11 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/FilmEquilateralTriangle.cs]: fixed

File: Task3/Task3/Task3/Figures/FilmSquare.cs (hunk 12 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/FilmSquare.cs]: +

File: Task3/Task3/Task3/Figures/FilmSquare.cs (hunk 13 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/FilmSquare.cs]: +

File: Task3/Task3/Task3/Figures/FilmSquare.cs (hunk 14 above)
Comment (Gaidukevich-Kirill) [reply Task3/Task3/Task3/Figures/FilmSquare.cs]: fixed

**Pull request merged**